### PR TITLE
Added diagnostic to print IOTA invariants H and I for all particles.

### PR DESCRIPTION
**Pull request by cemitch99**
Computes IOTA invariants of motion for all particles and prints to file.

1) Is the call from "DiagnosticOutput.cpp" the best place to put this?  

2) Input values of alpha, beta, t, and c are hard-coded, and this needs to be updated.

3) A follow-up request will update the IOTA thin lens element test problem to use this new diagnostic.

### CODE REVIEW

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 1]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+                    amrex::ParticleReal const cn = 0.01;
+                    Invariants Invariants(alpha,beta,tn,cn);
+
+                    // print out particles (this hack works only on CPU and on GPUs with
+                    // unified memory access)
+                    for (int i = 0; i < np; ++i) {
+
+                        // access AoS data such as positions and cpu/id
+                        PType const &p = aos_ptr[i];
+                        amrex::ParticleReal const x = p.pos(0);
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp:96]: As discussed: `.cpu()` and `.id()` need to be combined in AMReX to obtain a globally unique particle ID.

I will port the function `WarpXUtilIO::localIDtoGlobal` over from WarpX to ABLASTR:
- https://github.com/ECP-WarpX/WarpX/pull/3115
- https://github.com/ECP-WarpX/impactx/pull/135

Now, you can use:
```C++
#include <ablastr/particles/IndexHandling.H>

// ...
uint64_t const global_id = ablastr::particles::localIDtoGlobal(p.id(), p.cpu());
```

File: src/ImpactX.cpp [hunk 2]
```diff
@@ -100,5 +100,10 @@ namespace impactx
                                       diagnostics::OutputType::PrintParticles,
                                       "diags/output_beam.txt");
 
+        // print the two invariants H and I
+        diagnostics::DiagnosticOutput(*m_particle_container,
+                                      diagnostics::OutputType::PrintInvariants,
+                                      "diags/output_invariants.txt");
```
Comment (ax3l) [inline src/ImpactX.cpp]: Idea if we add more invariants later
```suggestion
                                      "diags/output_nonlinear_lens_invariants.txt");
```

File: src/particles/diagnostics/DiagnosticOutput.H [hunk 3]
```diff
@@ -18,7 +18,8 @@ namespace impactx::diagnostics
      */
     enum class OutputType
     {
-        PrintParticles ///< ASCII diagnostics, for small tests only
+        PrintParticles, ///< ASCII diagnostics, for small tests only
+        PrintInvariants ///< Also as above
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.H]:
```suggestion
        PrintNonlinearLensInvariants ///< ASCII diagnostics for the IOTA nonlinear lens, for small tests only
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 4]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
                                 << px << " " << py << " " << pt << "\n";
                     } // i=0...np
                 } // if( otype == OutputType::PrintParticles)
+                if (otype == OutputType::PrintInvariants) {
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp]:
```suggestion
                if (otype == OutputType::PrintNonlinearLensInvariants) {
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 5]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+
+                    amrex::ParticleReal const alpha = 0.0;
+                    amrex::ParticleReal const beta = 1.0;
+                    amrex::ParticleReal const tn = 0.4;
+                    amrex::ParticleReal const cn = 0.01;
+                    Invariants Invariants(alpha,beta,tn,cn);
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp]:
```suggestion
                    NonlinearLensInvariants nonlinear_lens_invariants(alpha,beta,tn,cn);
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 6]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+                        amrex::ParticleReal const x = p.pos(0);
+                        amrex::ParticleReal const y = p.pos(1);
+
+                        // access SoA Real data
+                        amrex::ParticleReal const px = part_px[i];
+                        amrex::ParticleReal const py = part_py[i];
+
+                        // write particle invariant data to file
+                        invariants_out HI_out;
+                        HI_out = Invariants(x, y, px, py);
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp]:
```suggestion
                        NonlinearLensInvariantsData HI_out = nonlinear_lens_invariants(x, y, px, py);
```
or
```suggestion
                        auto HI_out = nonlinear_lens_invariants(x, y, px, py);
```

File: src/particles/diagnostics/Invariants.H [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+#include <cmath>
+
+namespace impactx
+{
+    struct invariants_out {
+        amrex::ParticleReal H;
+        amrex::ParticleReal I;
+    };
+
+    struct Invariants
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
    struct NonlinearLensInvariants
```

File: src/particles/diagnostics/Invariants.H [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
+
+namespace impactx
+{
+    struct invariants_out {
+        amrex::ParticleReal H;
+        amrex::ParticleReal I;
+    };
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
    struct NonlinearLensInvariantsData {
        amrex::ParticleReal H; ///< description ...
        amrex::ParticleReal I; ///< description ...
    };
```

File: src/particles/diagnostics/Invariants.H [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            amrex::ParticleReal Hinv = Hpotential.m_real;
+            amrex::ParticleReal Iinv = Ipotential.m_real;
+
+            // compute invariants H and I
+            amrex::ParticleReal Jz = xn*pyn - yn*pxn;
+            Hinv = (pow(xn,2) + pow(yn,2) + pow(pxn,2) + pow(pyn,2))/2
+                 + m_tn*Hinv;
+            Iinv = pow(Jz,2) + pow(pxn,2) + pow(xn,2) + m_tn*Iinv;
+
+            invariants_out HI_out;
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
            NonlinearLensInvariantsData HI_out;
```

File: src/particles/diagnostics/Invariants.H [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+         *
+         * @param x particle position in x
+         * @param y particle position in y
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         */
+
+
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        invariants_out operator() (
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
        NonlinearLensInvariantsData operator() (
```

File: src/particles/diagnostics/Invariants.H [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        /** This is a nonlinear invariant functor, so that a variable of this type
+         *  can be used like a nonlinear invariant function.
+         *
+         * @param x particle position in x
+         * @param y particle position in y
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         */
+
+
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 12]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+                        PType const &p = aos_ptr[i];
+                        amrex::ParticleReal const x = p.pos(0);
+                        amrex::ParticleReal const y = p.pos(1);
+
+                        // access SoA Real data
+                        amrex::ParticleReal const px = part_px[i];
+                        amrex::ParticleReal const py = part_py[i];
+
+                        // write particle invariant data to file
+                        invariants_out HI_out;
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp]:
```suggestion
```

File: src/particles/diagnostics/Invariants.H [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+#include "particles/ImpactXParticleContainer.H"
+
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
+
+namespace impactx
```
Comment (ax3l) [inline src/particles/diagnostics/Invariants.H]:
```suggestion
namespace impactx::diagnostics
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 14]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+                    Invariants Invariants(alpha,beta,tn,cn);
+
+                    // print out particles (this hack works only on CPU and on GPUs with
+                    // unified memory access)
+                    for (int i = 0; i < np; ++i) {
+
+                        // access AoS data such as positions and cpu/id
+                        PType const &p = aos_ptr[i];
+                        amrex::ParticleReal const x = p.pos(0);
+                        amrex::ParticleReal const y = p.pos(1);
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp:97]:
```suggestion
                        amrex::ParticleReal const y = p.pos(1);
                        uint64_t const global_id = ablastr::particles::localIDtoGlobal(p.id(), p.cpu());
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 15]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
...
+
+                        // access SoA Real data
+                        amrex::ParticleReal const px = part_px[i];
+                        amrex::ParticleReal const py = part_py[i];
+
+                        // write particle invariant data to file
+                        invariants_out HI_out;
+                        HI_out = Invariants(x, y, px, py);
+                        amrex::AllPrintToFile(file_name)
+                                << HI_out.H << " " << HI_out.I << "\n";
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp:111]:
```suggestion
                                << global_id << " "
                                << HI_out.H << " " << HI_out.I << "\n";
```

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 16]
```diff
@@ -70,6 +71,35 @@ namespace impactx::diagnostics
                                 << px << " " << py << " " << pt << "\n";
                     } // i=0...np
                 } // if( otype == OutputType::PrintParticles)
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp:81]: Above are the lines `amrex::AllPrintToFile(file_name) << "id x y t px py pt\n";` which print the table header.

We need to update those as well depending on `otype`.

File: src/ImpactX.cpp [hunk 17]
```diff
@@ -100,5 +100,10 @@ namespace impactx
                                       diagnostics::OutputType::PrintParticles,
                                       "diags/output_beam.txt");
 
+        // print the two invariants H and I
+        diagnostics::DiagnosticOutput(*m_particle_container,
+                                      diagnostics::OutputType::PrintInvariants,
```
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
                                      diagnostics::OutputType::PrintNonlinearLensInvariants,
```

**Review by ax3l — COMMENTED** [review]
Code Review & Print Particle IDs

File: src/particles/diagnostics/DiagnosticOutput.cpp [hunk 18]
```diff
@@ -74,6 +79,36 @@ namespace impactx::diagnostics
...
+                else if (otype == OutputType::PrintNonlinearLensInvariants) {
+
+                    amrex::ParticleReal const alpha = 0.0;
+                    amrex::ParticleReal const beta = 1.0;
+                    amrex::ParticleReal const tn = 0.4;
+                    amrex::ParticleReal const cn = 0.01;
```
Comment (ax3l) [inline src/particles/diagnostics/DiagnosticOutput.cpp:87]: To be updated with `ParmParse` logic (and then documented in user-facing parameter docs).

File: src/particles/diagnostics/NonlinearLensInvariants.H [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** Independent coordinates
+         *
+         * Independent coordinates that are tracked as invariants of motion.
+         */
+        struct Data {
+            amrex::ParticleReal H; ///< first generalized coordinate
+            amrex::ParticleReal I; ///< second generalized coordinate
+        };
```
Comment (ax3l) [inline src/particles/diagnostics/NonlinearLensInvariants.H]: Please double check if the documentation is descriptive enough.

Comment (cemitch99) [reply src/particles/diagnostics/NonlinearLensInvariants.H]: I will modify this documentation and push a new commit soon.

**Review by ax3l — APPROVED** [review]
Splendid :)

**Pull request merged**